text/2018-01-14-skiplist.md

```diff
@@ -235,6 +251,41 @@ Interesting observations:
   124 ms and with `ConcurrentSkipListMap` it takes 110 ms. Now that's much closer to `SkipMap`
   and `std::map`.
 
+### The cost of reference counting in `Entry`
+
+When iterating over a skip list we use `Entry`s, which are essentially reference-counted
+pointers to skip list nodes. That means iterating over 100 elements incurs the
+cost of 200 atomic increments and 200 atomic decrements.
+
+Methods that insert, remove, or search for an element return `Entry`, which means
+they too incur some cost spent on incrementing and decrementing a node's reference count.
+
+The current skip list implementation doesn't provide alternative methods that avoid
+reference counting (i.e. avoid using `Entry`), but in the future we should discuss how
+to add them. Broadly speaking, there are two general alternatives to entries:
+clones and guards. Here's an illustration on the `SkipMap::get` method:
+
+```rust
+// Reference counting: return an `Entry`.
+//
+// This is the method signature we currently have.
+fn get(&self, k: &K) -> Option<Entry<K, V>>;
+
+// Alternative #1: return a clone of the element.
+//
+// This means we're paying the price of cloning, but that's
+// not a problem if cloning is cheap.
+fn get_clone(&self, k: &K) -> Option<V> where V: Clone;
+
+// Alternative #2: return a guarded reference to the element
+// keeping the thread pinned.
+//
+// The main drawback here is that the user must be careful and
+// not keep the guarda live for too long, or else garbage
+// collection will get stuck.
+fn get_guard(&self, k: &K) -> Option<Guard<K, V>>;
+```
```

Comment:
Could you please also list `with_element<F: FnOnce<&V>>(f: F)` under alternative 2? It's effectively the same as `get_guard`, but makes the guarded region more visible.